src/evict/evict_file.c

```diff
@@ -74,10 +77,34 @@ __wt_evict_file(WT_SESSION_IMPL *session, int syncop)
 		switch (syncop) {
 		case WT_SYNC_CLOSE:
 			/*
-			 * Evict the page.
+			 * We have exclusive access and we're closing the file,
+			 * evict the page.
 			 */
 			WT_ERR(__wt_evict(session, ref, 1));
 			break;
+		case WT_SYNC_CLOSE_SWEEP:
+			/*
+			 * The sweep server found a clean, idle file and it's
+			 * trying to close it. There's no exclusive lock on
+			 * anything as this can take a long time; if the file
+			 * becomes active, we abort the close process.
+			 *
+			 * Skip the root, we're not closing the file yet.
+			 */
+			if (__wt_ref_is_root(ref))
+				break;
+
+			/*
+			 * If re-opened and/or modified or we can't evict the
+			 * page for any reason, abort the close process.
+			 */
+			if (dhandle->session_inuse ||
+			    __wt_page_is_modified(page) ||
+			    !WT_ATOMIC_CAS4(
+			    ref->state, WT_REF_MEM, WT_REF_LOCKED))
+				WT_ERR(EBUSY);
+			WT_ERR(__wt_evict(session, ref, 0));
```

Comment:
I think this walk is going to give up fairly quickly on any file that was ever modified.  Evicting a page with any page-modified flag will dirty its parent page (and the btree handle), so when this walk visits the parent, it will give up with `EBUSY`.  Since the tree is now dirty, it won't be swept again until another checkpoint goes through.

Maybe we should only check whether leaf pages are dirty before giving up?  There is still the problem of requiring a checkpoint in a clean file before sweep can fully discard it, and the associated locking problems between the sweep server and checkpoints, but at least it could be incremental...

FTR, the reason we couldn't just bail out of the `WT_SYNC_DISCARD` walk is that we have no way to know that another thread is trying to open this tree, since requests will block on the dhandle writelock.  Maybe a workable approach would be to have readers increment something before waiting on the handle lock, so the `WT_SYNC_DISCARD` walk can exit early?